File: go-api/internal/service/csv.go
package service

import (
	"bytes"
	"context"
	"encoding/csv"
	"fmt"
	"io"
	"strconv"
	"strings"
	"time"

	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/user/application-tracker/go-api/internal/db"
)

// CSVHeaders defines the expected CSV column headers.
var CSVHeaders = []string{
	"companyName", "positionTitle", "status", "dateApplied",
	"companyUrl", "jobPostingUrl", "companyCareerUrl", "companyCategory",
	"skillsMatch", "jobSource", "salaryMin", "salaryMax",
	"coverLetterRequired", "offerDueDate", "specialRequirements", "notes",
}

// ImportResult holds the result of a CSV import operation.
type ImportResult struct {
	Imported int      `json:"imported"`
	Skipped  int      `json:"skipped"`
	Errors   []string `json:"errors"`
}

// ImportCSV parses and imports applications from a CSV reader.
func ImportCSV(ctx context.Context, pool *pgxpool.Pool, reader io.Reader) (ImportResult, error) {
	result := ImportResult{
		Errors: []string{},
	}

	r := csv.NewReader(reader)
	r.TrimLeadingSpace = true

	headers, err := r.Read()
	if err != nil {
		return result, fmt.Errorf("failed to read CSV headers: %w", err)
	}

	// Build header index map
	headerIdx := make(map[string]int)
	for i, h := range headers {
		headerIdx[strings.TrimSpace(h)] = i
	}

	// Validate required columns
	if _, ok := headerIdx["companyName"]; !ok {
		return result, fmt.Errorf("missing required column: companyName")
	}
	if _, ok := headerIdx["positionTitle"]; !ok {
		return result, fmt.Errorf("missing required column: positionTitle")
	}

	// Collect existing records from DB for duplicate detection
	existingURLs := make(map[string]bool)
	existingPairs := make(map[string]bool)
	allApps, err := db.GetAllApplications(ctx, pool)
	if err != nil {
		return result, fmt.Errorf("failed to load existing applications: %w", err)
	}
	for _, app := range allApps {
		if app.JobPostingURL.Valid && app.JobPostingURL.String != "" {
			existingURLs[app.JobPostingURL.String] = true
		} else {
			// Fallback key for apps without a job posting URL
			key := strings.ToLower(app.CompanyName) + "|" + strings.ToLower(app.PositionTitle)
			existingPairs[key] = true
		}
	}

	// Track intra-file duplicates
	seenURLs := make(map[string]bool)
	seenPairs := make(map[string]bool)

	rowNum := 1
	for {
		row, err := r.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			result.Errors = append(result.Errors, fmt.Sprintf("row %d: parse error: %v", rowNum, err))
			rowNum++
			continue
		}

		getField := func(name string) string {
			idx, ok := headerIdx[name]
			if !ok || idx >= len(row) {
				return ""
			}
			return strings.TrimSpace(row[idx])
		}

		companyName := getField("companyName")
		if companyName == "" {
			result.Errors = append(result.Errors, fmt.Sprintf("row %d: companyName is required", rowNum))
			rowNum++
			continue
		}

		positionTitle := getField("positionTitle")
		if positionTitle == "" {
			result.Errors = append(result.Errors, fmt.Sprintf("row %d: positionTitle is required", rowNum))
			rowNum++
			continue
		}

		jobPostingURL := getField("jobPostingUrl")

		// Duplicate detection: use jobPostingUrl if present, else fall back to (companyName, positionTitle)
		if jobPostingURL != "" {
			if existingURLs[jobPostingURL] || seenURLs[jobPostingURL] {
				result.Skipped++
				rowNum++
				continue
			}
			seenURLs[jobPostingURL] = true
		} else {
			pairKey := strings.ToLower(companyName) + "|" + strings.ToLower(positionTitle)
			if existingPairs[pairKey] || seenPairs[pairKey] {
				result.Skipped++
				rowNum++
				continue
			}
			seenPairs[pairKey] = true
		}

		status := getField("status")
		if status == "" {
			status = "unsubmitted"
		}

		dateAppliedStr := getField("dateApplied")
		var dateApplied *string
		if dateAppliedStr != "" {
			dateApplied = &dateAppliedStr
		}
		dateApplied = applyStatusDateConstraint(status, dateApplied)

		var jobPostingURLPtr *string
		if jobPostingURL != "" {
			jobPostingURLPtr = &jobPostingURL
		}
		companyURL := getField("companyUrl")
		var companyURLPtr *string
		if companyURL != "" {
			companyURLPtr = &companyURL
		}
		companyCareerURL := getField("companyCareerUrl")
		var companyCareerURLPtr *string
		if companyCareerURL != "" {
			companyCareerURLPtr = &companyCareerURL
		}
		companyCategory := getField("companyCategory")
		var companyCategoryPtr *string
		if companyCategory != "" {
			companyCategoryPtr = &companyCategory
		}
		var skillsMatchPtr *int32
		if s := getField("skillsMatch"); s != "" {
			if n, err := strconv.ParseInt(s, 10, 32); err == nil {
				v := int32(n)
				skillsMatchPtr = &v
			}
		}
		jobSource := getField("jobSource")
		var jobSourcePtr *string
		if jobSource != "" {
			jobSourcePtr = &jobSource
		}

		var salaryMin *int32
		if s := getField("salaryMin"); s != "" {
			if n, err := strconv.ParseInt(s, 10, 32); err == nil {
				v := int32(n)
				salaryMin = &v
			}
		}
		var salaryMax *int32
		if s := getField("salaryMax"); s != "" {
			if n, err := strconv.ParseInt(s, 10, 32); err == nil {
				v := int32(n)
				salaryMax = &v
			}
		}

		coverLetterRequired := false
		if clr := getField("coverLetterRequired"); clr == "true" || clr == "1" {
			coverLetterRequired = true
		}

		offerDueDate := getField("offerDueDate")
		var offerDueDatePtr *string
		if offerDueDate != "" {
			offerDueDatePtr = &offerDueDate
		}
		specialRequirements := getField("specialRequirements")
		var specialRequirementsPtr *string
		if specialRequirements != "" {
			specialRequirementsPtr = &specialRequirements
		}
		notes := getField("notes")
		var notesPtr *string
		if notes != "" {
			notesPtr = &notes
		}

		input := ApplicationInput{
			CompanyName:         companyName,
			PositionTitle:       positionTitle,
			Status:              status,
			DateApplied:         dateApplied,
			CompanyURL:          companyURLPtr,
			JobPostingURL:       jobPostingURLPtr,
			CompanyCareerURL:    companyCareerURLPtr,
			CompanyCategory:     companyCategoryPtr,
			SkillsMatch:         skillsMatchPtr,
			JobSource:           jobSourcePtr,
			SalaryMin:           salaryMin,
			SalaryMax:           salaryMax,
			CoverLetterRequired: coverLetterRequired,
			OfferDueDate:        offerDueDatePtr,
			SpecialRequirements: specialRequirementsPtr,
			Notes:               notesPtr,
		}

		if _, err := CreateApplication(ctx, pool, input); err != nil {
			result.Errors = append(result.Errors, fmt.Sprintf("row %d: %v", rowNum, err))
		} else {
			result.Imported++
			if jobPostingURL != "" {
				existingURLs[jobPostingURL] = true
			} else {
				pairKey := strings.ToLower(companyName) + "|" + strings.ToLower(positionTitle)
				existingPairs[pairKey] = true
			}
		}
		rowNum++
	}

	return result, nil
}

// ExportCSV exports all applications as a CSV byte slice.
func ExportCSV(ctx context.Context, pool *pgxpool.Pool) ([]byte, error) {
	apps, err := db.GetAllApplications(ctx, pool)
	if err != nil {
		return nil, err
	}

	var buf bytes.Buffer
	w := csv.NewWriter(&buf)

	// Write headers
	if err := w.Write(CSVHeaders); err != nil {
		return nil, err
	}

	for _, app := range apps {
		row := appToCSVRow(app)
		if err := w.Write(row); err != nil {
			return nil, err
		}
	}
	w.Flush()
	if err := w.Error(); err != nil {
		return nil, err
	}

	return buf.Bytes(), nil
}

// GetTemplate returns a CSV template with headers and one example row.
func GetTemplate() []byte {
	var buf bytes.Buffer
	w := csv.NewWriter(&buf)

	_ = w.Write(CSVHeaders)
	_ = w.Write([]string{
		"Acme Corp",
		"Software Engineer",
		"applied",
		time.Now().Format("2006-01-02"),
		"https://acme.com",
		"https://acme.com/jobs/123",
		"https://acme.com/careers",
		"enterprise-software",
		"4",
		"linkedin",
		"100000",
		"150000",
		"false",
		"",
		"",
		"",
	})
	w.Flush()
	return buf.Bytes()
}

// appToCSVRow converts a db.Application to a CSV row.
func appToCSVRow(app db.Application) []string {
	dateApplied := ""
	if app.DateApplied.Valid {
		dateApplied = app.DateApplied.Time.Format("2006-01-02")
	}

	companyURL := ""
	if app.CompanyURL.Valid {
		companyURL = app.CompanyURL.String
	}
	jobPostingURL := ""
	if app.JobPostingURL.Valid {
		jobPostingURL = app.JobPostingURL.String
	}
	companyCareerURL := ""
	if app.CompanyCareerURL.Valid {
		companyCareerURL = app.CompanyCareerURL.String
	}
	companyCategory := ""
	if app.CompanyCategory.Valid {
		companyCategory = app.CompanyCategory.String
	}
	skillsMatch := ""
	if app.SkillsMatch.Valid {
		skillsMatch = strconv.Itoa(int(app.SkillsMatch.Int32))
	}
	jobSource := ""
	if app.JobSource.Valid {
		jobSource = app.JobSource.String
	}
	salaryMin := ""
	if app.SalaryMin.Valid {
		salaryMin = strconv.Itoa(int(app.SalaryMin.Int32))
	}
	salaryMax := ""
	if app.SalaryMax.Valid {
		salaryMax = strconv.Itoa(int(app.SalaryMax.Int32))
	}
	offerDueDate := ""
	if app.OfferDueDate.Valid {
		offerDueDate = app.OfferDueDate.Time.Format("2006-01-02")
	}
	specialRequirements := ""
	if app.SpecialRequirements.Valid {
		specialRequirements = app.SpecialRequirements.String
	}
	notes := ""
	if app.Notes.Valid {
		notes = app.Notes.String
	}

	return []string{
		app.CompanyName,
		app.PositionTitle,
		app.Status,
		dateApplied,
		companyURL,
		jobPostingURL,
		companyCareerURL,
		companyCategory,
		skillsMatch,
		jobSource,
		salaryMin,
		salaryMax,
		strconv.FormatBool(app.CoverLetterRequired),
		offerDueDate,
		specialRequirements,
		notes,
	}
}

File: fastapi/src/services/csv.py
import csv
import io
from uuid import UUID

import asyncpg

from ..schemas import CsvRow, ImportError, ImportResult
from .history import build_description, record_history
from .shared import parse_date

CSV_COLUMNS = [
    "companyName",
    "positionTitle",
    "dateApplied",
    "status",
    "companyUrl",
    "jobPostingUrl",
    "companyCareerUrl",
    "companyCategory",
    "skillsMatch",
    "jobSource",
    "coverLetterRequired",
    "specialRequirements",
    "salaryMin",
    "salaryMax",
    "notes",
    "offerDueDate",
    "isArchived",
]


def get_sample_csv() -> str:
    header = ",".join(CSV_COLUMNS)
    example = (
        "Acme Corp,Software Engineer,2026-01-15,applied,"
        "https://acme.com,https://acme.com/jobs/123,https://acme.com/careers,"
        "ai,4,linkedin,false,Must have 3+ years React experience,80000,120000,Great company culture,,false"
    )
    return f"{header}\n{example}\n"


async def export_to_csv(pool: asyncpg.Pool) -> str:
    rows = await pool.fetch(
        "SELECT * FROM applications ORDER BY date_applied NULLS LAST"
    )

    buf = io.StringIO()
    writer = csv.DictWriter(buf, fieldnames=CSV_COLUMNS)
    writer.writeheader()

    for row in rows:
        writer.writerow(
            {
                "companyName": row["company_name"],
                "positionTitle": row["position_title"],
                "dateApplied": row["date_applied"].isoformat() if row["date_applied"] else "",
                "status": row["status"],
                "companyUrl": row["company_url"] or "",
                "jobPostingUrl": row["job_posting_url"] or "",
                "companyCareerUrl": row["company_career_url"] or "",
                "companyCategory": row["company_category"] or "",
                "skillsMatch": str(row["skills_match"]) if row["skills_match"] is not None else "",
                "jobSource": row["job_source"] or "",
                "coverLetterRequired": str(row["cover_letter_required"]).lower()
                if row["cover_letter_required"] is not None
                else "",
                "specialRequirements": row["special_requirements"] or "",
                "salaryMin": str(row["salary_min"]) if row["salary_min"] is not None else "",
                "salaryMax": str(row["salary_max"]) if row["salary_max"] is not None else "",
                "notes": row["notes"] or "",
                "offerDueDate": row["offer_due_date"].isoformat() if row["offer_due_date"] else "",
                "isArchived": str(row["is_archived"]).lower(),
            }
        )

    return buf.getvalue()


async def _get_existing_job_posting_urls(pool: asyncpg.Pool) -> set[str]:
    rows = await pool.fetch(
        "SELECT job_posting_url FROM applications WHERE job_posting_url IS NOT NULL"
    )
    return {row["job_posting_url"] for row in rows}


async def import_from_csv(pool: asyncpg.Pool, content: bytes) -> ImportResult:
    text = content.decode("utf-8")
    reader = csv.DictReader(io.StringIO(text))

    if reader.fieldnames is None:
        return ImportResult(
            imported=0,
            skipped=0,
            errors=[ImportError(row=1, message="Failed to parse CSV file")],
        )

    existing_urls = await _get_existing_job_posting_urls(pool)
    seen_urls: set[str] = set()
    imported = 0
    skipped = 0
    errors: list[ImportError] = []

    for i, raw_row in enumerate(reader):
        row_num = i + 2  # 1-based, +1 for header

        # Convert empty strings to None
        cleaned = {k: (v if v != "" else None) for k, v in raw_row.items()}

        try:
            row = CsvRow.model_validate(cleaned, by_alias=True)
        except Exception as exc:
            errors.append(ImportError(row=row_num, message=str(exc)))
            continue

        # Duplicate detection
        if row.job_posting_url:
            if row.job_posting_url in existing_urls or row.job_posting_url in seen_urls:
                skipped += 1
                continue
            seen_urls.add(row.job_posting_url)

        try:
            app_row = await pool.fetchrow(
                """
                INSERT INTO applications (
                    company_name, position_title, date_applied, status,
                    company_url, job_posting_url, company_career_url,
                    company_category, skills_match, job_source,
                    cover_letter_required, special_requirements,
                    salary_min, salary_max, notes, offer_due_date,
                    is_archived
                ) VALUES (
                    $1, $2, $3, $4::python_fastapi.application_status,
                    $5, $6, $7,
                    $8::python_fastapi.company_category, $9, $10::python_fastapi.job_source,
                    $11, $12,
                    $13, $14, $15, $16,
                    $17
                )
                RETURNING id
                """,
                row.company_name,
                row.position_title,
                parse_date(row.date_applied),
                row.status.value if row.status else "unsubmitted",
                row.company_url,
                row.job_posting_url,
                row.company_career_url,
                row.company_category.value if row.company_category else None,
                row.skills_match,
                row.job_source.value if row.job_source else None,
                row.cover_letter_required,
                row.special_requirements,
                row.salary_min,
                row.salary_max,
                row.notes,
                parse_date(row.offer_due_date),
                row.is_archived if row.is_archived is not None else False,
            )
            assert app_row is not None
            app_id: UUID = app_row["id"]
            await record_history(pool, app_id, build_description("Imported from CSV"))
            imported += 1
        except Exception as exc:
            errors.append(ImportError(row=row_num, message=f"Database error: {exc}"))

    return ImportResult(imported=imported, skipped=skipped, errors=errors)
